# The List Data Type


In [2]:
my_list_var =  [1, 2, 3]

In [7]:
type(my_list_var)

list

In [ ]:
student_names = ['Javokhir', ]

In [5]:
animals = ['cat', 'bat', 'horse', 'elephant']
animals

['cat', 'bat', 'horse', 'elephant']

In [6]:
my_list_with_mixed_data_type= ['hello', 3.1415, True, None, 42]
my_list_with_mixed_data_type

['hello', 3.1415, True, None, 42]

In [ ]:
spam = ['cat', 'bat', 'rat', 'elephant']

# Getting Individual Values in a List with Indexes


In [8]:
my_list_with_mixed_data_type[0]

'hello'

In [9]:
my_list_with_mixed_data_type[1]

3.1415

# Negative Indexes


- you can use any valid variable names for naming the list

In [18]:
any_name= ['hello', 3.1415, True, None, 42]

In [19]:
I = 2
print(any_name[I])

True


In [15]:
my_list_with_mixed_data_type[5]

IndexError: list index out of range

# Getting a List from Another List with Slices
- one integer index
- two integer slice


In [20]:
any_name

['hello', 3.1415, True, None, 42]

In [23]:
new_list = any_name[1:4] # slicing the list. first is inclusive
new_list

[3.1415, True, None]

In [24]:
any_name[:4]# from the beginning till inded 4

['hello', 3.1415, True, None]

In [25]:
any_name[2:] # from the index 2, till the end

[True, None, 42]

In [36]:
middle_point = len(any_name)//2 # pirovt_index variable to store the middle point of the list
any_name = any_name[middle_point:]
any_name

[None, 42]

In [42]:
animals

['cat', 'bat', 'horse', 'elephant']

In [43]:
len(animals)

4

In [44]:
middle_point = len(animals)//2
middle_point


2

In [45]:
animals[middle_point]

'horse'

# Getting a List’s Length with the len() Function
- depending on the data type meaning of operators can change
    - len()
    - *

In [34]:
2*3

6

In [35]:
'hello'*3

'hellohellohello'

In [29]:
len('message')

7

In [30]:
len(any_name)

3

In [31]:
any_name

[True, None, 42]

In [32]:
animals

['cat', 'bat', 'horse', 'elephant']

In [33]:
len(animals) 

4

# Changing Values in a List with Indexes


In [37]:
ta_students= ['Javokhir', 'Habibullo']
ta_students

['Javokhir', 'Habibullo']

In [38]:
ta_students[0] = 'Humoyun' # to change the value fo the list you can use indexing+assignment operator
ta_students

['Humoyun', 'Habibullo']

In [39]:
ta_students = ['TA1','TA2']
ta_students

['TA1', 'TA2']

In [40]:
ta_students[1]='TA3'
ta_students

['TA1', 'TA3']

In [41]:
ta_students[-1] = 'TA64'
ta_students

['TA1', 'TA64']

# List Concatenation and List Replication


In [46]:
TAs_of_001 = ['Javokhir','Habibullo']
TAs_of_002 = ['Asadbek','Shamsqamar']



In [49]:
'Hello' +' '+'Class'

'Hello Class'

In [50]:
s = 'Hello' +' '+'Class' # this is a concatation expression
print(s)

Hello Class


In [51]:
len(s)

11

In [47]:
TAs = TAs_of_001 + TAs_of_002  # concatematopm operator 
TAs

['Javokhir', 'Habibullo', 'Asadbek', 'Shamsqamar']

In [48]:
TAs

['Javokhir', 'Habibullo', 'Asadbek', 'Shamsqamar']

In [52]:
students_score = [ 70]
students_score

[70]

In [53]:
len(students_score)

1

In [55]:
# make stuents_score variable store every students score: average
students_score = [ 70] * 60
len(students_score)

60

# Removing Values from Lists with del Statements


In [63]:
contacts = ["Person A", "Person B", "Person X"] # image 1000x1000 pixes 10^6

In [64]:
del contacts[:2]
contacts

['Person X']

In [65]:
del contacts

In [66]:
contacts

NameError: name 'contacts' is not defined

In [58]:
del contacts[2] 
contacts

['Person A', 'Person B']

In [59]:
del contacts[0] 
contacts

['Person B']

# Working with Lists
- lists can help store a group of similar values in a better way
    - more organized
    - less duplicate

In [ ]:
catName1 = 'Zophie'
catName2 = 'Pooka'
catName3 = 'Simon'
catName4 = 'Lady Macbeth'
catName5 = 'Fat-tail'
catName6 = 'Miss Cleo'

In [ ]:
# see the example codes

# For loops with lists

In [ ]:
supplies = ['pens', 'staplers', 'flamethrowers', 'binders']
for i in range(len(supplies)):
    print('Index ' + str(i) + ' in supplies is: ' + supplies[i])

# Using Enumerate

# The in and not in Operators
- in checks if element is inside the list
- not in checks if the element is not inside the list

In [ ]:
'howdy' in ['hello', 'hi', 'howdy', 'heyas']

# Using the random.choice() and random.shuffle() Functions with Lists
- choice will randomly select index
- shuffle shuffles the order

#  Augmented Assignment Operators
- When assigning a value to a variable, we frequently use the variable itself. 
- augmented operators
    - `+`
    - `-`
    - `*`
    - `/`
    - `%`

# Methods


## finding a value in a list with the index() Method


In [ ]:
spam = ['hello', 'hi', 'howdy', 'heyas']

## Adding Values to Lists with the append() and insert() Methods


In [ ]:
spam = ['cat', 'dog', 'bat']

## Removing Values from Lists with the remove() Method
 

In [ ]:
spam = ['cat', 'bat', 'rat', 'elephant']

### del vs remove

## Sorting the Values in a List with the sort() Method


In [ ]:
##

## Reversing the Values in a List with the reverse() Method


## Example Program 
- write a code for randomly selecting a student from the list of students


In [ ]:
div001_student_names=[
    'Sirojiddinov Abdulaziz Ulugbek Ugli',
    'Abdukarimov Humoyunbek Olimjon Ugli',
    'Ochilov Shokhrukh Uchkun Ugli',
    'Abdurasulov Farrukhbek Ziyovuddin Ugli',
    'Chutboev Habibullo Islombek Ugli',
    'Ravshanbekov Khikmatullo Khayotbek Ugl',
    'Matyakubov Azizbek Yorkinbek Ugli',
    'Ilkhomov Akhrorbek Mardonbek Ugli',
    'Shabonov Abdullo Abbos Ugli',
    'Rasulov Elyor Dusyor Ugli',
    'Yuldoshev Javokhir Yusufjon Ugli',
    'Ghising Gyalchhen',
    'Khursanov Nazarbek Ulugbek Ugli',
    'Tologonova Datkaiym Temirbekovna',
    'Bayasgalan Delgerjargal',
    'Akhmadova Durdona Odiljon Kizi',
    'Olimjonov Diyorjon Sohibjon Ugli',
    'Akhmadova Dilnavoz Odiljon Kizi',
    'Bakhromov Rustam Nozim Ugli',
    'Egamkulova Makhliyo Abduraim Kizi',
    'Orifova Mokhinur Khakim Kizi',
    'Yusupov Murodilla Nizomiddin Ugli',
    'Avazkhonov Murtazokhon Yusufkhon Ugli',
    'Nosirkhonova Mubina Fakhriddin Kizi',
    'Umarova Mukaddaskhon Khursandjon Kizi',
    'Kodirov Muhammadkodir Shokir Ugli',
    'Shavkatov Mukhammadazam Shukhratbek Ug',
    'Tuychiev Mirjalol Kushmatjon Ugli',
    'Karimov Bekhruzbek Gafurjon Ugli',
    'Sultonov Begzod Azamat Ugli',
    'Khatamjonov Bekzodjon Bunyodjon Ugli',
    'Hamraev Sardor Khikmatovich',
    'Urdushev Sarvar Ulugbek Ugli',
    'Usmanov Sarvarbek Dilshod Ugli',
    'Hoshimov Shahriyor Ahmadjon Ugli',
    'Normamatova Shakhrizoda Gulomjon Kizi',
    'Ibodullaeva Shakhzoda Khamidjon Kizi',
    'Ozodov Sherzod Khamdam Ugli',
    'Norboeva Sevinch Akmal Kizi',
    'Yakhshimurodova Sevinchbonu Khudarga K',
    'Shomurodov Shohruh Olimjon Ugli',
    'Khaitov Shokhrukhjon Shukhrat Ugli',
    'Tadzhieva Amira',
    'Esanboev Abrorjon Akramjon Ugli',
    'Fayzullaev Asadbek Khayrullo Ugli',
    'Abduldaeva Aiana Ulanbekovna',
    'Saydullaeva Aziza Olimovna',
    'Akramov Akromjon Ikrom Ugli',
    'Abdulkhaev Akbarjon Akhmad Ugli',
    'Subkhiddinov Anvarbek Kambarjon Ugli',
    'Dushamboev Ogabek Nurmat Ugli',
    'Rayimkulov Otabek Ilkhomjon Ugli',
    'Vahobova Ugiloy Bakhtiyor Kizi',
    'Ergashev Ilhomjon Odiljon Ugli',
    'Pirimova Zarnigor Akhadovna',
    'Igamberdiev Javokhir Eldor Ugli',
    'Bakhronova Jasmina Orzumurodovna',
    'Khonimkulov Jaloliddin Ismatilla Ugli',
    'Uralova Zebo Bahodirjon Kizi',
    'Nakhalboev Juratbek Bakhodir Ugli',
    'Abdumukhiddinov Ziyodbek Ikboljon Ugli',
    'Bulturboev Fakhriddin Farkhod Ugli',
    'Pivavar Darya',
    'Siddikov Khusniddin Bekhzod Ugli'
 ]

In [ ]:
div002_student_names = [
    'Mukhammadov Muminjon Otajon Ugli',
    'Kutlimuratov Madrakhim Khushnud Ugli',
    'Akimov Sarvar Ravshan Ugli',
    'Akhmadjonov Samandar Anvarjon Ugli',
    'Shokirov Saidalo Nematillo Ugli',
    'Sharifov Azizbek Sharofovich',
    'Khudaynazarov Sunnatbek Bakhodir Ugli',
    'Carrasco Soriano Diana Laura',
    'Zokirov Diyorbek Botir Ugli',
    'Muhammadkhonov Mavlonkhon Muhriddinkho',
    'Vu Nguyen Bao Ngoc',
    'Murtozaev Sardor Muzaffar Ugli',
    'Rustamov Shamsiddin Azam Ugli',
    'Kakharova Sevinch Javlon Kizi',
    'Jabborov Sunnatjon Bakhtiyor Ugli',
    'Hamidkhonov Amirkhon Obidkhon Ugli',
    'Mustafakulov Azizbek Furkat Ugli',
    'Ziyaviddinov Ibrohim Ilhom Ugli',
    'Turaev Jahongir Bakhtiyorovich',
    'Ishonkulov Jakhongir Khusan Ugli',
    'Chinieva Shokhida Dilshodovna',
    'Khamdamov Kamronbek Laziz Ugli',
    'Tursunov Firdavs Farhodovich',
    'Ha Thi My Linh',
    'Kurbonova Nilufar Mizrobovna',
    'Orziev Diyorbek Matnazar Ugli',
    'Obidov Dilshod Khurshedovich',
    'Togaev Ravshan Khayritdin Ugli',
    'Shodmonov Rustam Bakhtiyor Ugli',
    'Khudoyberdiev Madaminbek Pulat Ugli',
    'Ilhomov Mansur Ilhom Ugli',
    'Kakhorjonov Mansurbek Tokhirjon Ugli',
    'Tsogoo Munkhzul',
    'Adhamov Muhammadaziz Gayratjon Ugli',
    'Koilybay Miras',
    'Muhammedov Mironshoh Mirkhon Ugli',
    'Abdurasulov Mirsaid Alimardonovich',
    'Nematjonov Bahodir Nurmukhamad Ugli',
    'Pak Georgiy',
    'Abduvokhidov Bekmurod Abdulkhodi Ugli',
    'Eshmirzaev Sarvar Raufjonovich',
    'Kosimov Shamskamar Jahongir Ugli',
    'Baratov Sokhibjon Khurshedovich',
    'Shokulov Shohruh Tuymurod Ugli',
    'Meliboev Shukurullo Fakhriddin Ugli',
    'Baktybek Uulu Arsen',
    'Shermirzaev Abbosbek Isroiljon Ugli',
    'Kiyomov Asadbek Kosimjon Ugli',
    'Mukhtorjonov Asilbek Murodjon Ugli',
    'Suiindik Akniyet',
    'Nabijonov Javokhir Pulatjon Ugli',
    'Nimatilaev Javokhir Izatillo Ugli',
    'Khursanaliev Javokhir Mukhiddin Ugli',
    'Abdulkhodiev Jobirkhon Rus Tamjon Ugli',
    'Malabon Julliane Samantha',
    'Chaw Thiri San',
    'Sobitkhonov Kudratkhon Odilkhon Ugli',
    'Bekzhanov Tariel Bekzhanovich',
    'Raykhonov Tilov Alisher Ugli',
    'Otamirzaev Fayozbek Farkhodjon Ugli',
    'Pirmatov Feruzbek Ermat Ugli',
    'Davlatov Hamidullo Nosirjon Ugli'
 ]

# The Tuple Data Type
- mutable vs immutable

# Converting Types with the list() and tuple() Functions


## References


In [ ]:
# simple data types

In [ ]:
# lists
spam = [0, 1, 2, 3, 4, 5]
cheese = spam # The reference is being copied, not the list.
cheese[1] = 'Hello!' # This changes the list value.

## Identity and the id() Function


In [10]:
id('Howdy') # The returned number will be different on your machine.

140404532918704

In [11]:
bacon = 'Hello'
id(bacon)

140404441907952

In [ ]:
bacon += ' world!' # A new string is made from 'Hello' and ' world!'.
id(bacon)

## Passing references

In [ ]:
def eggs(someParameter):
    someParameter.append('Hello')

spam = [1, 2, 3]
eggs(spam)
print(spam)

- how to prevent that?

## The copy Module’s copy()  Functions


# Homework 5

### Q0. for Q0, you write a short program
- define a function that 
    - takes two params as shown below
    - removes given subject from the list
    - returns the reduced list
- run the function definition
- call your defined function with the following arguments


In [ ]:
def drop_subject(all_subjects, name_of_the_subject_to_drop):
    # --write your code below this line--

    # remove the subject to be dropped from the list
    # return reduced list
    


In [ ]:
### Then call your defined function with the following arguments:
drop_subject(['SW','Korean Lang','Math','Business','Other'], 'SW')


### Q1. What is []?

### Q2. How would you assign the value 'hello' as the third value in a list stored in a variable named spam? (Assume spam contains [2, 4, 6, 8, 10].)




For the following three questions, let’s say spam contains the list ['a', 'b', 'c', 'd'].

### Q3. What does spam[int(int('3' * 2) // 11)] evaluate to?

### Q4. What does spam[-1] evaluate to?

### Q5. What does spam[:2] evaluate to?


For the following three questions, let’s say bacon contains the list [3.14, 'cat', 11, 'cat', True].

## Q6. What does bacon.index('cat') evaluate to?

## Q7. What does bacon.append(99) make the list value in bacon look like?

## Q8. What does bacon.remove('cat') make the list value in bacon look like?